### PR TITLE
fix(ACIR): correct brillig parameter slice length for dynamic arrays

### LINKED ISSUE #10197: Range end index out of range for slice
### Aim

This code:

```noir
fn main(cond: bool, index: u32) {
    let mut slice = &[(0, [1, 2]), (3, [4, 5])];
    if cond {
        slice = slice.push_back((6, [7, 8]));
    }
    slice[index] = (9, [10, 11]);
    let (value, new_slice) = slice.pop_front();
    println(value);
    println(new_slice);
}
```

with this Prover.toml:

```toml
cond = true
index = 1
```

(though if `cond` is false or the index is different it's the same)

leads to the following error (in the second println call):

```
The application panicked (crashed).
Message:  range end index 11 out of range for slice of length 8
Location: /Users/asterite/Projects/noir/acvm-repo/brillig_vm/src/lib.rs:340
```

~It's unclear if this is the same issue as #10170, maybe not.~ -> definitely not as in the code in the first comment here there are no `push/pop` operations.

It only happens with slices where element sizes are different (if we change the slice type to be `(Field, Field)` or `(Field, [Field])` then it doesn't crash).

### Expected Behavior

The code doesn't crash.

### Bug

It crashes.

### To Reproduce

1.
2.
3.
4.

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Project Impact

None

### Blocker Context

_No response_

### Nargo Version

_No response_

### NoirJS Version

_No response_

### Proving Backend Tooling & Version

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### LINKED ISSUE #10199: Slice with zero-sized element crashes the compiler
### Aim

This code:

```noir
fn main() {
    let _slice: [()] = &[];
}
```

leads to this panic, when doing `nargo compile`:

```
The application panicked (crashed).
Message:  attempt to calculate the remainder with a divisor of zero
Location: compiler/noirc_evaluator/src/ssa/validation/mod.rs:223
```

We are likely not accounting for the fact that a slice's flattened size can be zero.

### Expected Behavior

The program shouldn't crash.

### Bug

It crashes.

### To Reproduce

1.
2.
3.
4.

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Project Impact

None

### Blocker Context

_No response_

### Nargo Version

_No response_

### NoirJS Version

_No response_

### Proving Backend Tooling & Version

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Resolves #10197
Resolves #10199

## Summary

Please double chech the math as I think it's right but I'm not entirely sure. 

## Additional Context

Maybe we are not supposed to fix these issues right away, but doing so helps with not bumping into them again during auditing (with different code).

## Documentation

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/acir/call/mod.rs [hunk 1]
```diff
@@ -188,8 +188,20 @@ impl Context<'_> {
                         .get(&value_id)
                         .expect("ICE: Unknown slice input to brillig")
                     {
-                        AcirValue::DynamicArray(AcirDynamicArray { len, .. }) => *len,
-                        AcirValue::Array(array) => array.len(),
+                        AcirValue::DynamicArray(AcirDynamicArray { len, .. }) => {
+                            // len holds the flattened length of all elements in the slice,
+                            // so to get the no-flattened length we need to divide by the flattened
+                            // length of a single slice entry
+                            let sum: usize =
```
Comment (guipublic) [inline compiler/noirc_evaluator/src/acir/call/mod.rs]: small nit but you could sum as u32 and convert to usize only the sum (instead of all the inner values)

Comment (asterite) [reply compiler/noirc_evaluator/src/acir/call/mod.rs]: Done! Is that more efficient or idiomatic?

File: compiler/noirc_evaluator/src/acir/call/mod.rs [hunk 2]
```diff
@@ -188,8 +188,20 @@ impl Context<'_> {
...
+                                item_types.iter().map(|typ| typ.flattened_size() as usize).sum();
+                            if sum == 0 { 0 } else { *len / sum }
```
Comment (guipublic) [inline compiler/noirc_evaluator/src/acir/call/mod.rs]: not sure if we should return 0 or fail with an error if the sum is zero ?
do you have a test for this case?

Comment (asterite) [reply compiler/noirc_evaluator/src/acir/call/mod.rs]: No, it was just defensive programming. Let me check if I can reproduce this...

Comment (asterite) [reply compiler/noirc_evaluator/src/acir/call/mod.rs]: Ha, cool! I found a bug related to this (but somewhere else, it seems):

```noir
fn main() {
    let slice: [()] = &[];
    println(slice);
}
```

leads to:

```
The application panicked (crashed).
Message:  attempt to calculate the remainder with a divisor of zero
Location: compiler/noirc_evaluator/src/ssa/validation/mod.rs:223
```

So the total size can be zero, it seems. I'll report this but I think checking for zero might be okay here.

Comment (asterite) [reply compiler/noirc_evaluator/src/acir/call/mod.rs]: https://github.com/noir-lang/noir/issues/10199

I'll fix it in this PR as the fix is relatively small.

Comment (asterite) [reply compiler/noirc_evaluator/src/acir/call/mod.rs]: Done! This PR now fixes #10199. I also added a new test program which crashes if we don't take into account that the flattened size can be zero (for both Array and DynamicArray).

File: compiler/noirc_evaluator/src/acir/call/mod.rs (hunk 1 above)
Comment (guipublic) [reply compiler/noirc_evaluator/src/acir/call/mod.rs]: I'd say both, IMHO it's a bit easier to read because the closure is simpler, and we avoid many conversions so a bit more efficient. But yes it is a very small nit.
